File: components/deno/Infer.test.ts
import { assertEquals } from "https://deno.land/std@0.137.0/testing/asserts.ts";
import { Constraints } from "./Constraints.ts";
import { inferExpression } from "./Infer.ts";
import { parse } from "./Parser.ts";
import {
  createFresh,
  emptyTypeEnv,
  Scheme,
  TArr,
  TVar,
  Type,
  typeBool,
  typeInt,
} from "./Typing.ts";

const assertTypeEquals = (t: Type, expected: string) => {
  assertEquals(t.toString(), expected);
};

const assertConstraintsEquals = (
  constraints: Constraints,
  expected: Array<string>,
) => {
  assertEquals(constraints.toString(), expected.join(", "));
};

Deno.test("infer Apply", () => {
  const [constraints, type] = inferExpression(
    emptyTypeEnv
      .extend("a", new Scheme(["T"], new TArr(new TVar("T"), new TVar("T"))))
      .extend("b", new Scheme([], typeInt)),
    parse("a b"),
    new Constraints(),
    createFresh(),
  );

  assertConstraintsEquals(constraints, [
    "V1 -> V1 ~ Int -> V2",
  ]);
  assertTypeEquals(type, "V2");
});

Deno.test("infer If", () => {
  const [constraints, type] = inferExpression(
    emptyTypeEnv
      .extend("a", new Scheme(["S"], new TVar("S")))
      .extend("b", new Scheme([], typeInt))
      .extend("c", new Scheme(["T"], new TVar("T"))),
    parse("if (a) b else c"),
    new Constraints(),
    createFresh(),
  );

  assertConstraintsEquals(constraints, [
    "V1 ~ Bool",
    "Int ~ V2",
  ]);
  assertTypeEquals(type, "Int");
});

Deno.test("infer LBool", () => {
  const [constraints, type] = inferExpression(
    emptyTypeEnv,
    parse("True"),
    new Constraints(),
    createFresh(),
  );

  assertConstraintsEquals(constraints, []);
  assertTypeEquals(type, "Bool");
});

Deno.test("infer Lam", () => {
  const [constraints, type] = inferExpression(
    emptyTypeEnv,
    parse("\\x -> x 10"),
    new Constraints(),
    createFresh(),
  );

  assertConstraintsEquals(constraints, [
    "V1 ~ Int -> V2",
  ]);
  assertTypeEquals(type, "V1 -> V2");
});

Deno.test("infer Let", () => {
  const [constraints, type] = inferExpression(
    emptyTypeEnv,
    parse("let x = 10; y = x + 1 in y"),
    new Constraints(),
    createFresh(),
  );

  assertConstraintsEquals(constraints, [
    "Int -> Int -> V1 ~ Int -> Int -> Int",
  ]);
  assertTypeEquals(type, "Int");
});

Deno.test("infer LInt", () => {
  const [constraints, type] = inferExpression(
    emptyTypeEnv,
    parse("123"),
    new Constraints(),
    createFresh(),
  );

  assertEquals(constraints.constraints.length, 0);
  assertTypeEquals(type, "Int");
});

Deno.test("infer Op", () => {
  const scenario = (input: string, resultType: Type) => {
    const [constraints, type] = inferExpression(
      emptyTypeEnv
        .extend("a", new Scheme(["T"], new TVar("T")))
        .extend("b", new Scheme(["T"], new TVar("T"))),
      parse(input),
      new Constraints(),
      createFresh(),
    );

    assertConstraintsEquals(constraints, [
      `V1 -> V2 -> V3 ~ Int -> Int -> ${resultType}`,
    ]);
    assertTypeEquals(type, "V3");
  };

  scenario("a + b", typeInt);
  scenario("a - b", typeInt);
  scenario("a * b", typeInt);
  scenario("a / b", typeInt);
  scenario("a == b", typeBool);
});

Deno.test("infer Var", () => {
  const [constraints, type] = inferExpression(
    emptyTypeEnv.extend(
      "a",
      new Scheme(["T"], new TArr(new TVar("T"), new TVar("T"))),
    ),
    parse("a"),
    new Constraints(),
    createFresh(),
  );

  assertConstraintsEquals(constraints, []);
  assertTypeEquals(type, "V1 -> V1");
});

File: components/kotlin/app/src/test/kotlin/stlc/InferTest.kt
package stlc

import kotlin.test.Test
import kotlin.test.assertEquals

class InferTest {
    @Test
    fun inferApply() {
        val (constraints, type) = infer(
            emptyTypeEnv
                    + Pair("a", Scheme(setOf("T"), TArr(TVar("T"), TVar("T"))))
                    + Pair("b", Scheme(emptySet(), typeInt)),
            parse("a b")
        )

        assertConstraints(
            constraints, listOf(
                "V1 -> V1 ~ Int -> V2"
            )
        )
        assertEquals(TVar("V2"), type)
    }

    @Test
    fun inferIf() {
        val (constraints, type) = infer(
            emptyTypeEnv
                    + Pair("a", Scheme(setOf("S"), TVar("S")))
                    + Pair("b", Scheme(emptySet(), typeInt))
                    + Pair("c", Scheme(setOf("S"), TVar("S"))),
            parse("if (a) b else c")
        )

        assertConstraints(
            constraints, listOf(
                "V1 ~ Bool",
                "Int ~ V2"
            )
        )
        assertEquals(typeInt, type)
    }

    @Test
    fun inferLam() {
        val (constraints, type) = infer(
            emptyTypeEnv,
            parse("\\x -> x 10")
        )

        assertConstraints(
            constraints, listOf(
                "V1 ~ Int -> V2"
            )
        )
        assertEquals("V1 -> V2", type.toString())
    }

    @Test
    fun inferLBool() {
        val (constraints, type) = infer(emptyTypeEnv, parse("True"))

        assertConstraints(constraints, emptyList())
        assertEquals(typeBool, type)
    }

    @Test
    fun inferLInt() {
        val (constraints, type) = infer(emptyTypeEnv, parse("123"))

        assertConstraints(constraints, emptyList())
        assertEquals(typeInt, type)
    }

    @Test
    fun inferLet() {
        val (constraints, type) = infer(emptyTypeEnv, parse("let x = 10; y = x + 1 in y"))

        assertConstraints(
            constraints, listOf(
                "Int -> Int -> V1 ~ Int -> Int -> Int"
            )
        )
        assertEquals(typeInt, type)
    }

    @Test
    fun inferOp() {
        fun scenario(input: String, resultType: Type) {
            val (constraints, type) = infer(
                emptyTypeEnv
                        + Pair("a", Scheme(setOf("T"), TVar("T")))
                        + Pair("b", Scheme(setOf("T"), TVar("T"))),
                parse(input)
            )

            assertConstraints(
                constraints, listOf(
                    "V1 -> V2 -> V3 ~ Int -> Int -> $resultType"
                )
            )

            assertEquals(TVar("V3"), type)
        }

        scenario("a + b", typeInt)
        scenario("a - b", typeInt)
        scenario("a * b", typeInt)
        scenario("a / b", typeInt)
        scenario("a == b", typeBool)
    }

    @Test
    fun inferVar() {
        val (constraints, type) = infer(
            emptyTypeEnv
                    + Pair("a", Scheme(setOf("T"), TArr(TVar("T"), TVar("T")))), parse("a")
        )

        assertEquals(constraints, Constraints())
        assertEquals(TArr(TVar("V1"), TVar("V1")), type)
    }

    private fun assertConstraints(constraints: Constraints, expected: List<String>) {
        assertEquals(constraints.toString(), expected.joinToString(", "))
    }
}
